src/data/num/to_bits.lean

```diff
@@ -0,0 +1,150 @@
+  equiv_list_bool.symm (tt :: l) = bit1 (equiv_list_bool.symm l) := rfl
+@[simp] lemma equiv_list_bool_symm_cons_ff (l : list bool) :
+  equiv_list_bool.symm (ff :: l) = bit0 (equiv_list_bool.symm l) := rfl
+
+lemma to_trailing_bits_len (n : pos_num) : (equiv_list_bool n).length = nat.log 2 n :=
+begin
+  induction n with b ih b ih, { erw equiv_list_bool_one, simp, },
+  { simp [cast_bit1, nat.bit1_val b, ih], },
+  { simp [cast_bit0, nat.bit0_val b, ih], }
+end
+
+lemma equiv_list_bool_symm_lt (ls : list bool) : (equiv_list_bool.symm ls : ℕ) < 2^(ls.length+1) :=
+begin
+  induction ls with hd tl ih, { simp [equiv_list_bool], },
+  cases hd,
+  { simp only [equiv_list_bool_symm_cons_ff, pos_num.cast_bit0],
+    rw [nat.bit0_val, mul_comm, pow_add], simpa, },
+  { simp only [equiv_list_bool_symm_cons_tt, pos_num.cast_bit1],
+    rw [nat.bit1_val, add_comm, pow_add],
+    apply @nat.lt_of_div_lt_div _ _ 2,
+    simpa [nat.add_mul_div_left, nat.div_eq_of_lt], },
+end
+
+end pos_num
+
+namespace num
+
+/-- An encoding function of the binary numbers in bool. -/
+def to_bits : num → list bool
+| num.zero := []
+| (num.pos n) := (pos_num.equiv_list_bool n) ++ [tt]
+
+@[simp] lemma to_bits_zero : to_bits 0 = [] := rfl
+
+@[simp] lemma to_bits_bit0 (n : num) (hn : n ≠ 0) : (num.bit0 n).to_bits = ff :: n.to_bits :=
+by { cases n, { contradiction, }, simp [to_bits, num.bit0], }
+
+@[simp] lemma to_bits_bit1 (n : num) : (num.bit1 n).to_bits = tt :: n.to_bits :=
+by { cases n, { refl, }, simp [to_bits, num.bit1], }
+
+/-- A decoding function from `list bool` to `num` -/
+def of_bits (L : list bool) : num := if L = [] then 0 else pos_num.equiv_list_bool.symm L.init
+
+@[simp] lemma of_bits_nil : of_bits [] = 0 := rfl
+@[simp] lemma of_bits_cons_ff (l : list bool) (hl : l ≠ []) :
+  of_bits (ff :: l) = num.bit0 (of_bits l) := by simp [of_bits, hl, num.bit0]
+@[simp] lemma of_bits_cons_tt (l : list bool) : of_bits (tt :: l) = num.bit1 (of_bits l) :=
+by { cases l, { refl, }, simp [of_bits, num.bit1], }
+
+@[simp] lemma of_bits_to_bits (n : num) : of_bits (to_bits n) = n :=
+by { cases n, { refl, }, simp [of_bits, to_bits], }
+
+lemma of_bits_to_bits_left_inv : function.left_inverse of_bits to_bits := of_bits_to_bits
+
+lemma mem_to_bits_range_iff (l : list bool) :
+  l ∈ set.range to_bits ↔ l = [] ∨ tt ∈ l.last' :=
+begin
+  split, { rintro ⟨n, h⟩, subst h, cases n; simp [to_bits], },
+  apply list.reverse_cases_on l, { intro, use 0, refl, },
+  intros st lt h,
+  use (pos_num.equiv_list_bool.symm st),
+  symmetry, simpa [to_bits] using h,
+end
+
+@[simp] lemma to_bits_of_bits_valid {l : list bool} (h : l ∈ set.range to_bits) :
+  to_bits (of_bits l) = l := of_bits_to_bits_left_inv.right_inv_on_range h
+
+@[simp] lemma of_bits_invalid (l : list bool) :
+  of_bits (l ++ [ff]) = of_bits (l ++ [tt]) := by simp [of_bits]
+
+lemma to_bits_len (n : num) : (to_bits n).length = if n = 0 then 0 else nat.log 2 n + 1 :=
```

Comment:
This should be stated with `nat.size` (either as a replacement or addition)

Reply:
Based on Zulip, I would put this in a separate PR that shows that it is equal to `nat.bits`